### PR TITLE
feat: Invite link - create invite links for organisations

### LINKED ISSUE #102: [Feat] Create An Invite to Join An Organization
## Description
Create a system for generating invitations for users to join organizations.

## Authentication and Authorization
* API endpoint requires a valid JWT token in the Authorization header.
* Only owners for the specific organization can create invitations.

## Acceptance Criteria
* Owners can create an invitation for an email address.
* Owners can specify which organisation a user can be invited to.
* Invitations should have an expiration date.

## Data Validation and Sanitization:
* The API should validate the request payload to ensure the email field is present and valid.
* The API should validate that the org_id is a valid UUID.
* The JWT in the Authorization header must be valid and not expired.
* The API should validate that the requesting user has ownership rights for the specified organization.

## Requirements
* Create API for generating an invitation for an email address.
* Include JWT authentication protection for the API endpoint.
* Implement authorization checks for ownership.
* Write tests for the invitation system

## Purpose
* To create an invite for a user to join an organisation.

## Expected Outcome
* A fully functional invitation system that allows an organisation owner to create invitations for new members or users to join organisations.

## API Endpoints
1. Create Invitation(s) [POST] /api/v1/organisations/create-invite
**Request Headers:** Authorization: Bearer <JWT_TOKEN>
**Request Body:**
```
{
  "email": "user1@example.com",
  "org_id": "a1b2c3d4-e5f6-7890-abcd-ef1234567890"
}
```

**Success Response (201 Created):**
Content-Type: application/json
```
{
  "status": "success"
  "message": "Invitation created successfully",
  "data": {
    "invitation": {
      "invite_id": "string",
      "email": "user1@example.com",
      "organization_id": "string",
      "invite_link": "string" 
      "status": "pending",
      "expires_at": "2023-07-23T00:00:00Z"
    }
  }
}
```

**Failure Response (422 Bad Request):**
Content-Type: application/json
```
{
 "errors": [
    {
      "field": "string",
      "message": "string"
    },
  ]
}
```

**Failure Response (403 Forbidden):**
Content-Type: application/json
```
{
  "error": "Unauthorized",
  "message": "You do not have ownership rights for this organization"
}
```
**Failure Response (404 Not Found):**
Content-Type: application/json
```
{
  "error": "Not Found",
  "message": "No organization was found for the given details"
}
```

## Testing
* Unit tests for invitation creation
* Integration tests for the end-to-end invitation flow
* Test cases for all error scenarios
* Test cases for authorization and authentication

### PR DESCRIPTION
**Pull request by Ace2489**
# Description

Closes #102

# Changes proposed
## What were you told to do?
Implement the organization invite feature, allowing users to invite others to join their organization.

## What did you do?
- Created a new `OrganizationInvite` entity in the `Domain` layer
- Implemented `IOrganizationInviteService` and `OrganizationInviteService` in the `Infrastructure` layer
- Added `CreateOrganizationInviteCommand` and `CreateOrganizationInviteCommandHandler` in the `Application` layer
- Created `OrganizationInviteDto` and `CreateOrganizationInviteDto` in the `Application` layer
- Implemented `OrganizationInviteMapperProfile` for AutoMapper
- Added a new API endpoint in `OrganizationController` for creating invites
- Implemented unit tests for the new command handler and service
- Updated the database context to include the new `OrganizationInvite` entity
- Added necessary validations for email and organization existence

# Check List (Check all the applicable boxes)
🚨Please review the [contribution guideline](CONTRIBUTING.md) for this repository.

- [x] My code follows the code style of this project.
- [x] This PR does not contain plagiarized content.
- [x] The title and description of the PR is clear and explains the approach.
- [x] I am making a pull request against the **dev branch** (left side).
- [x] My commit messages styles matches our requested structure.
- [x] My code additions will fail neither code linting checks nor unit test.
- [x] I am only making changes to files I was requested to.

# Screenshots/Videos

## New API Endpoint
`POST /api/organizations/invite`

### Request Body
```json
{
  "email": "newuser@example.com",
  "userId": "00000000-0000-0000-0000-000000000000"
}
```

### Response (Success)
```json
{
  "id": "11111111-1111-1111-1111-111111111111",
  "organizationId": "22222222-2222-2222-2222-222222222222",
  "email": "newuser@example.com",
  "status": "Pending",
  "inviteLink": "https://example.com/invite/abc123",
  "expiresAt": "2023-07-31T23:59:59Z"
}
```

### CODE REVIEW

**Comment by chimaihueze** [comment]
You have conflicting files. please resolve.

**Review by NdubuisiJr — CHANGES_REQUESTED** [review]
Fix the two comments below and your merge conflict

File: src/Hng.Infrastructure/Repository/OrganizationInviteRepository.cs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+using Hng.Domain.Entities;
+using Hng.Infrastructure.Context;
+
+namespace Hng.Infrastructure.Repository.Interface;
+
+public class OrganizationInviteRepository(ApplicationDbContext context) : Repository<OrganizationInvite>(context), IOrganizationInviteRepository
```
Comment (NdubuisiJr) [inline src/Hng.Infrastructure/Repository/OrganizationInviteRepository.cs:6]: You don't need this. You didn't even use it. Delete everything relating to it.

File: src/Hng.Infrastructure/Services/OrganizationInviteService.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
+using System.Security.Cryptography;
+using System.Text;
+using Hng.Domain.Entities;
+using Hng.Infrastructure.Repository.Interface;
+using Hng.Infrastructure.Services.Interfaces;
+
+namespace Hng.Infrastructure.Services;
+
+public class OrganizationInviteService(IRepository<Organization> Organizationrepository, IOrganizationInviteRepository repository) : IOrganizationInviteService
+{
```
Comment (NdubuisiJr) [inline src/Hng.Infrastructure/Services/OrganizationInviteService.cs:10]: Remove everything concerning `IOrganizationInviteRepository`. Use `IRepository<OrganisationInvite`, that will give you all the methods to access your table.

File: src/Hng.Web/Controllers/OrganizationController.cs [hunk 3]
```diff
@@ -35,4 +39,24 @@ public async Task<ActionResult<OrganizationDto>> CreateOrganization([FromBody] C
...
+        body.UserId = inviterId;
+        var command = new CreateOrganizationInviteCommand(body);
+        var response = await mediator.Send(command);
+
+        if (response == null)
+        {
+            return UnprocessableEntity();
+        }
+
+        return this.CreatedResult("Invitation created successfully", new { invitation = response });
```
Comment (NdubuisiJr) [inline src/Hng.Web/Controllers/OrganizationController.cs:60]: Remove the `this` keyword.

**Review by NdubuisiJr — CHANGES_REQUESTED** [review]

**Pull request closed without merging**